hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -523,15 +523,15 @@ void updateCurrBlockRef(HFileBlock block) {
       if (block != null && curBlock != null && block.getOffset() == curBlock.getOffset()) {
         return;
       }
-      if (this.curBlock != null) {
+      if (this.curBlock != null && !this.curBlock.isOnHeap()) {
         prevBlocks.add(this.curBlock);
       }
       this.curBlock = block;
     }
 
     void reset() {
-      // We don't have to keep ref to EXCLUSIVE type of block
-      if (this.curBlock != null) {
+      // We don't have to keep ref to heap block
+      if (this.curBlock != null && !this.curBlock.isOnHeap()) {
```

Comment:
Ya here only difference is you will see whether the block is a Shared memory backed. If so add ref. We dont really worry whether it is off heap or on heap.

Reply:
Adding to prevBlocks is only at these 2 places?

Reply:
> Adding to prevBlocks is only at these 2 places?

Yeah, check all the places,  only the two add here.

Reply:
Fine